modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/config/DeferredConfigurationProvider.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.extension.internal.runtime.config;
+
+import static org.mule.runtime.core.api.util.ClassUtils.withContextClassLoader;
+import static org.mule.runtime.module.extension.internal.util.MuleExtensionUtils.getClassLoader;
+
+import org.mule.runtime.api.component.AbstractComponent;
+import org.mule.runtime.api.event.Event;
+import org.mule.runtime.api.meta.model.ComponentModel;
+import org.mule.runtime.api.meta.model.ExtensionModel;
+import org.mule.runtime.api.meta.model.config.ConfigurationModel;
+import org.mule.runtime.core.api.el.ExpressionManager;
+import org.mule.runtime.core.api.event.CoreEvent;
+import org.mule.runtime.extension.api.runtime.config.ConfigurationInstance;
+import org.mule.runtime.extension.api.runtime.config.ConfigurationProvider;
+import org.mule.runtime.module.extension.internal.runtime.resolver.ValueResolver;
+import org.mule.runtime.module.extension.internal.runtime.resolver.ValueResolvingContext;
+
+/**
+ * A {@link ConfigurationProvider} that defers the resolution of the actual {@link ConfigurationProvider} to be used until an
+ * event is received.
+ *
+ * @since 4.5.0
+ */
+public final class DeferredConfigurationProvider extends AbstractComponent implements ConfigurationProvider {
+
+  /**
+   *
+   * @param extensionModel the model that owns the {@link ConfigurationModel}.
+   * @param operationModel the model for the component that requires the configuration.
+   * @return A {@link ConfigurationModel} suitable for the given {@link ComponentModel}.
+   */
+  private static ConfigurationModel getConfigurationModelForComponent(ExtensionModel extensionModel,
+                                                                      ComponentModel operationModel) {
+    return extensionModel.getConfigurationModels().stream()
+        .filter(cm -> cm.getOperationModels().stream().anyMatch(om -> om.getName().equals(operationModel.getName())))
+        .findAny()
+        .orElse(null);
+  }
```

Comment:
Is this really valid?

Reply:
i don't get the purpose of this. The same operation can be compatible with N different configs. Why is this relevant at all since at the end of the day it's the user the one that should be resolving the config through the expression?

Reply:
I did this because the `ConfigurationProvider` has a method for getting the `ConfigurationModel`. That method is being called during initialization for validation purposes (more specifically [here](https://github.com/mulesoft/mule/blob/aa340afc2ec146c7e1f5706bf2285bf76f9d4269/modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/OperationMessageProcessor.java#L134)).
I wasn't sure whether a config-ref for an operation could be compatible with configs of different `ConfigurationModels`.
But now I should see if I can do something else about this. Probably that validation may have to be skipped at that stage, if an expression was used.

Reply:
THat's a good point, but you won't be able to do that validation anyway right? Also, the `getConfigurationModel()` implementation that you have in this class is not accurate as you have no way of being sure if that's the configuraitonModel that you'll get. What it's clear though is that the validation that you pointed out should only apply for static config-refs.

Reply:
The implementation has changed. We are using a `ValueResolver` now, which does not require implementing `getConfigurationModel`.
The validation mentioned in the discussion can now be skipped if the configuration is dynamic.